File: AILang/benchmarks/c/file_io.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>

#define PAYLOAD "abcdefghijklmnopqrstuvwxyz0123456789"


int32_t file_io_bench(int32_t iterations) {
    const char* path = "benchmarks/out/file_io_bench.txt";
    const size_t payload_len = (size_t)strlen(PAYLOAD);
    int64_t checksum = 0;
    char buffer[256];

    for (int32_t i = 0; i < iterations; i++) {
        FILE* out = fopen(path, "wb");
        if (out == NULL) {
            return -1;
        }
        fwrite(PAYLOAD, 1, payload_len, out);
        fclose(out);

        FILE* in = fopen(path, "rb");
        if (in == NULL) {
            return -1;
        }
        size_t read_len = fread(buffer, 1, sizeof(buffer), in);
        (void)fclose(in);
        checksum += (int64_t)read_len;
    }

    return (int32_t)checksum;
}

int main(void) {
    int32_t iterations = 2000;
    int64_t result = file_io_bench(iterations);
    printf("%lld\n", (long long)result);
    return 0;
}

File: AILang/benchmarks/rust/file_io.rs
use std::fs;

fn file_io_bench(iterations: i32) -> i64 {
    let path = "benchmarks/out/file_io_bench.txt";
    let payload = "abcdefghijklmnopqrstuvwxyz0123456789";
    let mut checksum: i64 = 0;

    for _ in 0..iterations {
        fs::write(path, payload).expect("write file");
        let content = fs::read_to_string(path).expect("read file");
        checksum += content.len() as i64;
    }
    checksum
}

fn main() {
    let iterations = 2000;
    let result = file_io_bench(iterations);
    println!("{}", result);
}
